File: sdk/typescript/src/pkce.ts
/**
 * PKCE (Proof Key for Code Exchange) Utilities
 */

/**
 * Generate random code verifier
 */
export function generateCodeVerifier(): string {
  const array = new Uint8Array(32);
  crypto.getRandomValues(array);
  return base64UrlEncode(array);
}

/**
 * Generate code challenge from verifier using S256 method
 */
export async function generateCodeChallenge(verifier: string): Promise<string> {
  const encoder = new TextEncoder();
  const data = encoder.encode(verifier);
  const hash = await crypto.subtle.digest('SHA-256', data);
  return base64UrlEncode(new Uint8Array(hash));
}

/**
 * Generate both verifier and challenge
 */
export async function generatePKCEPair(): Promise<{
  verifier: string;
  challenge: string;
}> {
  const verifier = generateCodeVerifier();
  const challenge = await generateCodeChallenge(verifier);
  return { verifier, challenge };
}

/**
 * Base64 URL encode without padding
 */
function base64UrlEncode(buffer: Uint8Array): string {
  const base64 = btoa(String.fromCharCode(...buffer));
  return base64
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=/g, '');
}

/**
 * Generate random state parameter
 */
export function generateState(): string {
  return generateCodeVerifier();
}

File: packages/sdk-python/oauth_mcp_gateway/pkce.py
"""
PKCE (Proof Key for Code Exchange) Utilities
"""

import secrets
import hashlib
import base64
from typing import Tuple


def generate_code_verifier() -> str:
    """Generate random code verifier"""
    return base64.urlsafe_b64encode(secrets.token_bytes(32)).decode("utf-8").rstrip("=")


def generate_code_challenge(verifier: str) -> str:
    """Generate code challenge from verifier using S256 method"""
    digest = hashlib.sha256(verifier.encode("utf-8")).digest()
    return base64.urlsafe_b64encode(digest).decode("utf-8").rstrip("=")


def generate_pkce_pair() -> Tuple[str, str]:
    """Generate both verifier and challenge"""
    verifier = generate_code_verifier()
    challenge = generate_code_challenge(verifier)
    return verifier, challenge


def generate_state() -> str:
    """Generate random state parameter"""
    return generate_code_verifier()
